libbeat/esleg/eslegclient/connection.go

```diff
@@ -184,15 +184,18 @@ func NewConnection(s ConnectionSettings) (*Connection, error) {
 		logger.Info("kerberos client created")
 	}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
+	// There are some cases where the connection is created but Connect
+	// is not called before it's used, so we populate reqsContext and cancelReqs
+	// here.
```

Comment:
Can we just find the places where we don't call connect and fix them? This code is only called within Beats and we can search for uses of `esleg`.

Also using `context.Background()` is a smell, the parent context should be an argument, which will have the compiler find all NewConnection uses for you so you can audit them to see if they use close inappropriately.

The connection here also don't really represent a network connection at the network level, it looks like it is a convenience wrapper around an HTTP client. From that perspective closing or having a connection level context doesn't make much sense, it is just a wrapper for closing idle connections.

Arguably the contexts should be set on a per request basis in all the places that call `execRequest` and then the cancellation should propagate through those individual calls.

Reply:
> Can we just find the places where we don't call connect and fix them? This code is only called within Beats and we can search for uses of esleg.

I went with this approach to be on the safe side of not introducing the possibility of a panic for calling a `nil` function. The panic I got when running the tests seems to be coming form a Connection that is not used but it's closed. That is triggered by a test testing a failure scenario where the ES host is not reachable.

Anyways, I've been looking into that.

> Arguably the contexts should be set on a per request basis in all the places that call execRequest and then the cancellation should propagate through those individual calls.

I'm not sure that would achieve the same effect we currently have. Currently `reqsContext` is used to cancel in-flight requests when the Connection needs to be shutdown, that is done by the `Close` method that is called by a different gorotine than the one waiting for the in-flight request(s) to finish.

The issues that are fixed by this new behaviour:
- https://github.com/elastic/beats/issues/40518
- https://github.com/elastic/beats/issues/38666

An the PR fixing them:
- https://github.com/elastic/beats/pull/40572


Honestly, I rather have this PR merged as is, so the issues listed above and https://github.com/elastic/beats/issues/40705 are correctly fixed on `main`. Then we can create a follow up issue to refactor the code ensuring that:
- Methods creating requests accept a context
- In-flight requests can be cancelled when the Connection is closed by a different goroutine.

Reply:
I finally figured out the reason why Filebeat would panic when the connection was closed. It's an interesting corner case.
The root cause is if the publishing pipeline never tries to publish an event and Filebeat is shutdown, in that case the connection to ES was never used, however it is closed during the shutdown process, leading to a panic if `cancelReqs` is nil.

I could add some checks to ensure `cancelReqs` and `reqsContext` are not used if `nil`, however it feels cleaner to keep the previous behaviour of `NewConnection` returning a Connection that is safe to use without any change of behaviour on its methods.